src/Fabulous/Program.fs

```diff
@@ -58,23 +72,44 @@ module Program =
           Logger = ProgramDefaults.defaultLogger()
           ExceptionHandler = ProgramDefaults.defaultExceptionHandler }
 
+    //TODO when would I use this one? How does it compare to the other stateful* builders?
```

Comment:
`stateful` is the simplest type of program where a state is used. You send a message, the state is updated.

`statefulWithCmd` adds the concept of side effects on top of the state via `Cmd`s

`statefulWithCmdMsg` is the same than `statefulWithCmd` except it adds a layer of discriminated union to be easier to unit test. You can read more about it on https://zaid-ajaj.github.io/the-elmish-book/#/chapters/scaling/intent (Intent is just a different naming for the same concept)

Reply:
For `'arg`, please see explanation above

Reply:
@TimLariviere I've read about _Intent_ and understand that it's a pattern used in component programs for cleanly separating messages to the component from messages to the parent program. It's used to avoid leaking the decision into the parent program about which component messages to intercept, inspect or forward. The parent only needs to worry about handling the intents and can blindly pass component messages on to the component's update function.

The diff of https://github.com/fabulous-dev/Fabulous.Avalonia/pull/224, which switches the examples between between using `statefulWithCmdMsg` and `statefulWithCmd`, looks like what I've tried to describe [here](https://github.com/fabulous-dev/Fabulous.Avalonia/pull/224#issuecomment-1999759296).

I'm failing to see the connection between _Intent_ and _CmdMsg_ - other than there's command mapping and batching involved. Can you help me connect the dots?

Reply:
@h0lg True, sorry about the confusion. `Intent` is indeed another concept we called `ExternalMsg` in Fabulous to enable child -> parent communication. Not the same thing at all than `CmdMsg`.

Your understanding of `CmdMsg` is correct.

When side-effects are required, `init` and `update` can return a `Cmd<'msg>` along the updated `Model`. This `Cmd<'msg>` is essentially a list of functions to be executed by Fabulous as side-effects, which might dispatch new messages.

The nice thing with MVU is that it's very easy to unit test since the `init` and `update` functions are pure and will always return an updated state, so you can check that a specific message with a specific state will return the expected updated state.

But unit testing the returned `Cmd<'msg>` is extremely difficult as you would need to execute it yourself. What if that `Cmd<'msg>` is making a network call or database update?

It's easier to separate the "intention" (not `Intent`, that's where I got confused earlier) of side-effects with the actual implementation as it can just check against the "intention" (the `CmdMsg`) in the unit tests.

```fs
type Msg = | Clicked

let doNetworkCallCmd () =
    Cmd.OfAsync...

let update msg model =
    match msg with
    | Clicked -> { model with Clicked = true }, doNetworkCallCmd()

let ``When user clicks the button, do the network call``() =
    let initialState = { Clicked = false }
    let expectedState = { Clicked = true }

    let actualState, actualCmd = update Clicked initialState

    Assert.Equals(actualState, expectedState)
    // TODO: How do you test that "actualCmd" is doing the right thing without actually doing the network call?
```

```fs
type Msg = | Clicked

type CmdMsg = | DoNetworkCall

let doNetworkCallCmd () =
    Cmd.OfAsync...

let mapCmdMsgToMsg cmdMsg =
    match cmdMsg with
    | DoNetworkCall -> doNetworkCallCmd()

let update msg model =
    match msg with
    | Clicked -> { model with Clicked = true }, [ DoNetworkCall ]

let ``When user clicks the button, do the network call``() =
    let initialState = { Clicked = false }
    let expectedState = { Clicked = true }
    let expectedCmdMsg = DoNetworkCall

    let actualState, actualCmdMsg = update Clicked initialState

    Assert.Equals(actualState, expectedState)
    Assert.Equals(actualCmdMsg, expectedCmdMsg) // Way easier to unit test
```

Reply:
@TimLariviere I think I understand now - and thank you for helping me!

`statefulWithCmdMsg` allows you to refactor side-effects out of your `update` function, making it pure and easy to unit-test without executing the side-effects.

You do so by defining an extra discriminated union for messages that trigger side-effects (in your example called `CmdMsg`) and then return those messages from your update function instead of executing the side effects in place. 

An additional function (in your example called `mapCmdMsgToMsg`) maps each of those messages back to the side-effect.

If I understand correctly, I suggest that besides updating the code doco (which I'll try), your example would be a good addition to the doco as a topic called _Unit testing the update function_ or something similar.

To
- make this concept more accessible to newbies by getting closer to known ELM/MVU concepts,
- avoid confusion between `CmdMsg`, `Cmd<'msg>` and `ExternalMsg` and
- capture the variant of Program this creates,

I suggest the following renaming:

`CmdMsg` -> `SideEffect` (in examples - of course we can call it whatever we like)
`mapCmdMsgToMsg` -> `mapSideEffectToMsg` or simply `mapSideEffect`
`statefulWithCmdMsg` -> `statefulWithSideEffects` or `statefulWithIsolatedSideEffects`

Reply:
@TimLariviere I have updated the doco on the stateful* builders according to my new found understanding - thanks again for clarifying!